### DISCUSSION TITLE
Proposal for Tiny Site URL Shortener Project

### DISCUSSION THREAD

**Discussion by sahsisunny** [Real-Dev-Squad/website-todo-items#176] [General]
## RFC 
- https://github.com/Real-Dev-Squad/todo-action-items/issues/175

## Description:
This discussion thread serves as a proposal for the development of a Tiny Site URL Shortener project. The purpose of this proposal is to outline the objectives, scope, technical requirements, and key features of the project. We invite all contributors, and interested parties to review and provide feedback on the proposal before proceeding with the development.

## Key Points to Discuss:

- **Project Objectives:** Share your thoughts on the outlined objectives of the Tiny Site URL Shortener project. Are there any additional objectives that should be considered?
- **Scope and Features:** Provide feedback on the scope of the project and the proposed features. Are there any features you believe should be added or modified?
- **Technical Requirements:** Discuss the listed technical requirements and suggest any alternative technologies or approaches that might enhance the project's implementation.
- **User Experience:** Share your insights on the user interface and experience. Are there any design considerations or usability aspects that should be addressed?

Your feedback is crucial in shaping the direction of this project. Let's collaborate to create an efficient and user-friendly Tiny Site URL Shortener that meets the needs of our users. Please feel free to share your thoughts, ideas, and any other relevant considerations.

**Comment by j24m** [comment]
This is my version of the RFC :

# Tiny Site(URL Shorteners) :

## **Project Information:**

- **Project Name** :  Tiny Site (URL shorteners)
- **Version** : 1

## Introduction :

- **Overview** : Tiny Site is a user-friendly website that reduces the length of your URL(Uniform Resource Locator). The idea is to minimize the web page address into something that’s easier to remember and track. 
This web application will maintain connection between the new URL and the original URL, meaning that anyone who clicks on your new URL will be redirected to the original page.
- Scope : pending
- Intended Audience : pending
- **Context and Motivation for developing the app** : 
→ Make URLs user friendly and memorable. 
→ Enable users to track and analyze URL statistics, including access frequency and timestamps of accesses.
→ Decrease the character numbers for social media sharing.

## Requirement :

- Given a long URL, the service should generate a shorter and unique alias for it.
- When the user hits a short link, the service should redirect to the original link.
- Links will expire after a standard default time span.
- The system should be highly available. This is really important to consider because if the service goes down, all the URL redirection will start failing.
- URL redirection should happen in real-time with minimal latency.
- Shortened links should not be predictable.

## System Architecture :

- Frontend: A user interface for submitting and managing URLs.
- Backend: Responsible for URL shortening, redirect handling, and analytics processing.
- Database: Store mappings between shortened and original URLs, as well as usage statistics.

## **Technical Requirements**

- Programming Language: Choose a suitable programming language for the backend and frontend development.
- Frameworks: Select relevant frameworks for web development and database interaction.
- Data Storage: Implement a database system for storing URL mappings and analytics data.

## **Functional Requirements**

### → **Shortening**

- Users can submit a long URL through a user interface.
- The backend generates a unique short code for the URL and stores the mapping.
- The short URL is returned to the user for sharing.

### → **Redirecting**

- When a user accesses a shortened URL, the application retrieves the corresponding long URL from the database.
- The user is then redirected to the original long URL.

### **→ Analytics (optional)**

- Usage statistics are collected for each shortened URL, including the number of clicks, timestamp, and geographic location.
- Analytics data can be visualized through an admin interface.

## **User Interface Design**

The user interface shall include:

- A form for submitting long URLs and obtaining shortened URLs.
- A page for displaying usage statistics and analytics (optional).

## **Performance and Scalability**

- Design the system to handle a large number of URL shortening requests and redirections.
- Implement caching mechanisms to improve response times.

## **Security and Privacy**

- Implement secure practices to prevent unauthorized access to the URL mapping database.
- Consider using encryption for sensitive data.
- Provide clear information to users regarding data collection and usage (if analytics are enabled).

## **Testing and Quality Assurance**

- Develop unit tests, integration tests, and user acceptance tests to ensure the application functions as intended.

## **Deployment and Release**

- Specify the deployment process, hosting environment, and version control strategy.
